src/devices/machine/z80sio.cpp

```diff
@@ -1891,37 +1905,57 @@ void z80sio_channel::advance_rx_fifo()
 			m_rx_error_fifo >>= 8;
 
 			// load error status from the FIFO
-			m_rr1 = (m_rr1 & ~m_rr1_auto_reset) | uint8_t(m_rx_error_fifo & 0x000000ffU);
-
-			// if we're in interrupt-on-first mode, clear interrupt if there's no pending error condition
-			if ((m_wr1 & WR1_RX_INT_MODE_MASK) == WR1_RX_INT_FIRST)
-			{
-				for (int i = 0; m_rx_fifo_depth > i; ++i)
-				{
-					if (uint8_t(m_rx_error_fifo >> (i * 8)) & (RR1_CRC_FRAMING_ERROR | RR1_RX_OVERRUN_ERROR))
-						return;
-				}
-				m_uart->clear_interrupt(m_index, INT_RECEIVE);
-			}
+			m_rr1 = (m_rr1 & ~m_rr1_auto_reset) | uint8_t(m_rx_error_fifo & 0x000000ffU & ~RR1_HIDDEN_1ST_MARKER);
 		}
 		else
 		{
 			// no more characters available in the FIFO
 			m_rr0 &= ~RR0_RX_CHAR_AVAILABLE;
-			if ((m_wr1 & WR1_WRDY_ENABLE) && (m_wr1 & WR1_WRDY_ON_RX_TX))
-				set_ready(false);
-			m_uart->clear_interrupt(m_index, INT_RECEIVE);
+			update_wait_ready();
 		}
 	}
+	update_rx_int();
 }
 
 uint8_t z80sio_channel::get_special_rx_mask() const
 {
-	return ((m_wr4 & WR4_STOP_BITS_MASK) == WR4_STOP_BITS_SYNC) ?
-		(RR1_RX_OVERRUN_ERROR | RR1_END_OF_FRAME) :
-		(RR1_RX_OVERRUN_ERROR | RR1_CRC_FRAMING_ERROR);
+	if ((m_wr1 & WR1_RX_INT_MODE_MASK) == WR1_RX_INT_DISABLE)
+		return 0;
+	else
+	{
+		uint8_t mask = ((m_wr4 & WR4_STOP_BITS_MASK) == WR4_STOP_BITS_SYNC) ?
+			(RR1_RX_OVERRUN_ERROR | RR1_END_OF_FRAME) :
+			(RR1_RX_OVERRUN_ERROR | RR1_CRC_FRAMING_ERROR);
+		if ((m_wr1 & WR1_RX_INT_MODE_MASK) == WR1_RX_INT_ALL_PARITY)
+			mask |= RR1_PARITY_ERROR;
+		return mask;
+	}
 }
 
+void z80sio_channel::update_rx_int()
+{
+	bool state = false;
+
+	auto rx_int_mode = m_wr1 & WR1_RX_INT_MODE_MASK;
+	if (rx_int_mode != WR1_RX_INT_DISABLE)
+	{
+		if (m_rr1 & get_special_rx_mask())
+			state = true;
+		else if (m_rx_fifo_depth)
+		{
+			// FIFO not empty
+			if (rx_int_mode != WR1_RX_INT_FIRST)
+				state = true;
+			else if (m_rx_error_fifo & RR1_HIDDEN_1ST_MARKER)
+				state = true;
+		}
+	}
+	LOGINT("rx %d wr1 %02x rr1 %02x fd %u ref %06x\n", state, m_wr1, m_rr1, m_rx_fifo_depth, m_rx_error_fifo);
+	if (state)
+		m_uart->trigger_interrupt(m_index, INT_RECEIVE);
+	else
+		m_uart->clear_interrupt(m_index, INT_RECEIVE);
+}
```

Comment:
Ok, I added checks in `trigger_interrupt` and `clear_interrupt` that cause `check_interrupt` to be called only in case the interrupt request changes state.